google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPreconditionsSupport.java

```diff
@@ -0,0 +1,199 @@
+    final CloudSdkService.SdkStatusUpdateListener sdkStatusUpdateListener =
+        (sdkService, status) -> {
+          switch (status) {
+            case READY:
+            case INVALID:
+            case NOT_AVAILABLE:
+              installationCompletionLatch.countDown();
+              break;
+            case INSTALLING:
+              // continue waiting for completion.
+              break;
+          }
+        };
+
+    if (installInProgress) {
+      cloudSdkService.addStatusUpdateListener(sdkStatusUpdateListener);
+      deploymentLog.print(GctBundle.getString("appengine.deployment.status.preparing.sdk") + "\n");
+    } else {
+      // no need to wait for install if unsupported or completed.
+      installationCompletionLatch.countDown();
+    }
+
+    // wait for SDK to be ready and trigger the actual deployment if it properly installs.
+    ProgressManager.getInstance()
+        .run(
+            new Task.Backgroundable(
+                project, GctBundle.message("appengine.deployment.status.deploying"), true) {
+              @Override
+              public void run(@NotNull ProgressIndicator indicator) {
+                try {
+                  while (installationCompletionLatch.getCount() > 0) {
+                    // wait interruptibility to check for user cancel each second.
+                    installationCompletionLatch.await(1, SECONDS);
+                    if (checkIfCancelled()) {
+                      break;
+                    }
+                  }
+
+                  // at this point the installation should be either ready or user cancelled.
+                  ApplicationManager.getApplication()
+                      .invokeLater(
+                          () -> {
+                            doRun(cloudSdkService, runnable, callback);
+                          });
+
+                } catch (InterruptedException e) {
+                  /* valid cancellation exception, no handling needed. */
+                } finally {
+                  // remove the notification listener regardless of waiting outcome.
+                  cloudSdkService.removeStatusUpdateListener(sdkStatusUpdateListener);
+                }
+              }
+            });
+  }
+
+  @VisibleForTesting
+  boolean checkIfCancelled() {
+    return ProgressManager.getInstance().getProgressIndicator().isCanceled();
+  }
+
+  private void doRun(
+      CloudSdkService cloudSdkService, Runnable runnable, DeploymentOperationCallback callback) {
+    // check the status of SDK after install.
+    SdkStatus postInstallSdkStatus = cloudSdkService.getStatus();
+    switch (postInstallSdkStatus) {
+      case INSTALLING:
+        String message = GctBundle.message("appengine.deployment.error.sdk.still.installing");
+        callback.errorOccurred(message);
+        showCloudSdkNotification(
+            message, NotificationType.WARNING, false /* no settings needed for this case. */);
+        return;
+      case NOT_AVAILABLE:
```

Comment:
For now it means it was never installed or failed to install / was cancelled. For custom path SDK it means the path is empty. Managed SDK does not have `INVALID` status per se, since we always attempt to re-install if it's not `READY`.